### PR TITLE
src: Add vk_format_utils to UtilityHeaders

### PR DESCRIPTION
**Pull request by charles-lunarg**
The library originates from Vulkan-ValidationLayers, but is being moved into this repo to make it easier for others to use it.

The library has also been modified to be header only and C compatible, which allows more developers to be able to use it. This does require some changes, but only affects the FormatElementSize and FormatTexelSize functions which used default parameters. Two new functions, FormatElementSizeWithAspect and FormatTexelSizeWithAspect have been added to handle the non-default image aspect case (the default was COLOR_BIT).

### CODE REVIEW

File: tests/add_subdirectory/vk_format_utils.c [hunk 1]
```diff
@@ -0,0 +1,8 @@
+// Copyright 2023 The Khronos Group Inc.
+// Copyright 2023 Valve Corporation
+// Copyright 2023 LunarG, Inc.
+//
+// SPDX-License-Identifier: Apache-2.0
+#include <vulkan/utility/vk_format_utils.h>
+
+void foo() {}
```
Comment (spencer-lunarg) [inline tests/add_subdirectory/vk_format_utils.c:8]: what is this for?

File: include/vulkan/utility/vk_format_utils.h [hunk 2]
```diff
@@ -0,0 +1,2297 @@
...
+    FORMAT_UTILS_NUMERICAL_TYPE_SNORM,
+    FORMAT_UTILS_NUMERICAL_TYPE_SRGB,
+    FORMAT_UTILS_NUMERICAL_TYPE_SSCALED,
+    FORMAT_UTILS_NUMERICAL_TYPE_UFLOAT,
+    FORMAT_UTILS_NUMERICAL_TYPE_UINT,
+    FORMAT_UTILS_NUMERICAL_TYPE_UNORM,
+    FORMAT_UTILS_NUMERICAL_TYPE_USCALED
+};
+
+enum FORMAT_UTILS_COMPATIBILITY_CLASS {
```
Comment (spencer-lunarg) [inline include/vulkan/utility/vk_format_utils.h:35]: 1. would it be better to forward declare so people looking at the file can quickly see the functions at the top?
2. I guess curious if this should be `VK_FORMAT_UTILS_*`... should we use `VK_FORMAT_UTILS` name space to match the filename

**Review by spencer-lunarg — CHANGES_REQUESTED** [review]
I guess before we merge here, a good exercise would be go in the Validation Layers

1. Remove the current format utils (so it doesn't generate)
2. Replace the header with this file
3. Make sure CI will pass and this didn't forget anything subtle a test there would have caught

(the goal not being to merge that PR, but to prove this new interface works)

File: tests/add_subdirectory/vk_format_utils.c (hunk 1 above)
Comment (charles-lunarg) [reply tests/add_subdirectory/vk_format_utils.c:8]: This tests whether the code is actually valid C code - since I just need to include it for the compiler to check, I should probably rename the function and add some usage just for testings sake.

File: include/vulkan/utility/vk_format_utils.h (hunk 2 above)
Comment (charles-lunarg) [reply include/vulkan/utility/vk_format_utils.h:35]: 1. Forward declaring the enums is a good idea, I made the functions all forward declared for the same logic after all. 

2. VK_ looks like a vulkan api enum which I do not want to be mistaken as.

**Comment by charles-lunarg** [comment]
> I guess before we merge here, a good exercise would be go in the Validation Layers

I started to do that on Friday and did not have much trouble with the conversion but had a lot of trouble running tests (w and w/o mock icd). It felt like I setup the repo wrong but I was on a different machine than I normally develop on, so I’ll try again come the morn on that machine.

File: tests/add_subdirectory/vk_format_utils.c (hunk 1 above)
Comment (artem-lunarg) [reply tests/add_subdirectory/vk_format_utils.c:8]: Is it possible to make this file "ephemeral", by generating only during the build phase? If that makes sense in this context. I guess this method is used by some configuration tools that generate small files to check for support of particular platform/language feature.

File: include/vulkan/utility/vk_format_utils.h [hunk 3]
```diff
@@ -0,0 +1,2297 @@
...
+    FORMAT_UTILS_COMPATIBILITY_CLASS_PVRTC2_2BPP,
+    FORMAT_UTILS_COMPATIBILITY_CLASS_PVRTC2_4BPP,
+    FORMAT_UTILS_COMPATIBILITY_CLASS_S8
+};
+// Return the plane index of a given VkImageAspectFlagBits.
+//     VK_IMAGE_ASPECT_PLANE_0_BIT -> 0
+//     VK_IMAGE_ASPECT_PLANE_1_BIT -> 1
+//     VK_IMAGE_ASPECT_PLANE_2_BIT -> 2
+//     <any other value> -> FORMAT_UTILS_MAX_PLANES
+inline uint32_t GetPlaneIndex(VkImageAspectFlagBits aspect);
```
Comment (artem-lunarg) [inline include/vulkan/utility/vk_format_utils.h:126]: One option is to return uint_max here (and to rename constant to something with INVALID or similar). Invalid program has more chances not to crash with value 3 (which is close to correct values) than with a large value. Another consideration point is to which degree should we care about backward compatibility. Using large invalid constant will help the library to be less affected by some future PLANE_3_BIT (not sure how probably is it).

File: include/vulkan/utility/vk_format_utils.h [hunk 4]
```diff
@@ -0,0 +1,2297 @@
...
+    FORMAT_UTILS_NUMERICAL_TYPE_NONE = 0,
+    FORMAT_UTILS_NUMERICAL_TYPE_SFLOAT,
+    FORMAT_UTILS_NUMERICAL_TYPE_SINT,
+    FORMAT_UTILS_NUMERICAL_TYPE_SNORM,
+    FORMAT_UTILS_NUMERICAL_TYPE_SRGB,
+    FORMAT_UTILS_NUMERICAL_TYPE_SSCALED,
+    FORMAT_UTILS_NUMERICAL_TYPE_UFLOAT,
+    FORMAT_UTILS_NUMERICAL_TYPE_UINT,
+    FORMAT_UTILS_NUMERICAL_TYPE_UNORM,
+    FORMAT_UTILS_NUMERICAL_TYPE_USCALED
```
Comment (artem-lunarg) [inline include/vulkan/utility/vk_format_utils.h:32]: nit: suggestion to end last enum line with a comma, to remove corresponding line from the future diffs if new members are added.

File: include/vulkan/utility/vk_format_utils.h [hunk 5]
```diff
@@ -0,0 +1,2297 @@
...
+    FORMAT_UTILS_COMPATIBILITY_CLASS_S8
+};
+// Return the plane index of a given VkImageAspectFlagBits.
+//     VK_IMAGE_ASPECT_PLANE_0_BIT -> 0
+//     VK_IMAGE_ASPECT_PLANE_1_BIT -> 1
+//     VK_IMAGE_ASPECT_PLANE_2_BIT -> 2
+//     <any other value> -> FORMAT_UTILS_MAX_PLANES
+inline uint32_t GetPlaneIndex(VkImageAspectFlagBits aspect);
+
+// Returns whether a format is of the numberical type SFLOAT
```
Comment (artem-lunarg) [inline include/vulkan/utility/vk_format_utils.h:128]: typo: numberical

File: include/vulkan/utility/vk_format_utils.h [hunk 6]
```diff
@@ -0,0 +1,2297 @@
...
+
+inline bool FormatsEqualComponentBits(VkFormat format_a, VkFormat format_b) {
+    const struct FORMAT_UTILS_FORMAT_INFO format_info_a = GetFormatInfo(format_a);
+    const struct FORMAT_UTILS_FORMAT_INFO format_info_b = GetFormatInfo(format_b);
+    if (format_info_a.compatibility == FORMAT_UTILS_COMPATIBILITY_CLASS_NONE || format_info_b.compatibility == FORMAT_UTILS_COMPATIBILITY_CLASS_NONE) {
+        return false;
+    } else if (format_info_a.component_count != format_info_b.component_count) {
+        return false;
+    }
+    // Need to loop match each component type is found in both formats
```
Comment (artem-lunarg) [inline include/vulkan/utility/vk_format_utils.h:2259]: Why don't we match just corresponding components? Matching based on possible permutations is not intuitive and is not something I'd would expect from api like this. Matching based on direct permutations looks like more practical use case.

File: include/vulkan/utility/vk_format_utils.h [hunk 7]
```diff
@@ -0,0 +1,2297 @@
...
+// Returns whether a format is of the numberical type UNORM
+// Format must only contain one numberical type, so formats like D16_UNORM_S8_UINT always return false
+inline bool FormatIsUNORM(VkFormat format);
+
+// Returns whether a format is of the numberical type USCALED
+// Format must only contain one numberical type, so formats like D16_UNORM_S8_UINT always return false
+inline bool FormatIsUSCALED(VkFormat format);
+
+// Returns whether a format type is any type from "Interpretation of Numeric Format" table (OpTypeInt)
+inline bool FormatIsSampledInt(VkFormat format);
```
[the remark this answers is not in the document]
Comment (charles-lunarg) [reply include/vulkan/utility/vk_format_utils.h:165]: I wasn't super clear on that myself - but I'll look at the format table in question and double check that.

File: include/vulkan/utility/vk_format_utils.h [hunk 8]
```diff
@@ -0,0 +1,2297 @@
...
+
+inline bool FormatHasRed(VkFormat format) { return FormatHasComponentType(format, FORMAT_UTILS_COMPONENT_TYPE_R); }
+
+inline bool FormatHasGreen(VkFormat format) { return FormatHasComponentType(format, FORMAT_UTILS_COMPONENT_TYPE_G); }
+
+inline bool FormatHasBlue(VkFormat format) { return FormatHasComponentType(format, FORMAT_UTILS_COMPONENT_TYPE_B); }
+
+inline bool FormatHasAlpha(VkFormat format) { return FormatHasComponentType(format, FORMAT_UTILS_COMPONENT_TYPE_A); }
+
+inline bool FormatsEqualComponentBits(VkFormat format_a, VkFormat format_b) {
```
[the remark this answers is not in the document]
Comment (charles-lunarg) [reply include/vulkan/utility/vk_format_utils.h:2251]: FormatsIsEqualComponentBits sets off all sorts of 'grammar' warning bells in my head, as it should be FormatsAreEqualComponentBits... (singular vs plural).

Naming ~~are~~ is hard!

[the remark this answers is not in the document]
Comment (artem-lunarg) [reply include/vulkan/utility/vk_format_utils.h:2251]: 💯

File: include/vulkan/utility/vk_format_utils.h (hunk 6 above)
Comment (charles-lunarg) [reply include/vulkan/utility/vk_format_utils.h:2259]: So I didn't write the code, I just fixed it up for this PR. 

My understanding is that this is directly checking permutations of formats. R8G8B8A8 == B8G8R8A8 would be an example. Checking that R8G8B8A8_UNORM == R8G8B8A8_SRGB would require a different function, and I'm not sure vk_format_utils.h has anything for that currently.

File: include/vulkan/utility/vk_format_utils.h (hunk 7 above)
[the remark this answers is not in the document]
Comment (spencer-lunarg) [reply include/vulkan/utility/vk_format_utils.h:165]: so for the formats, awhile ago I tried to unify the wording

https://registry.khronos.org/vulkan/specs/1.3-extensions/html/vkspec.html#formats-numericformat

[internal MR](https://gitlab.khronos.org/vulkan/vulkan/-/merge_requests/5627)

in short:

- **Numeric Format** = [`SFLOAT`, `SINT`, `SNORM`, `SRGB`, `SSCALED`, `UFLOAT`, `UINT`, `UNORM`, `USCALED`]
- **Numeric Type** = [float, signed int, unsigned int]
- **SPIR-V Type** = [ [`OpTypeFloat`, `OpTypeInt`], width, sign ]

.... basically we should re-write this comment to whatever @charles-lunarg  thinks "makes sense" as I am now to tainted with spec lawyer knowledge

File: include/vulkan/utility/vk_format_utils.h (hunk 8 above)
[the remark this answers is not in the document]
Comment (spencer-lunarg) [reply include/vulkan/utility/vk_format_utils.h:2251]: I don't care what it is called, just that everything else is also consistently the same

> Naming is hard

:100:

File: include/vulkan/utility/vk_format_utils.h (hunk 6 above)
Comment (spencer-lunarg) [reply include/vulkan/utility/vk_format_utils.h:2259]: the only use of this is in `VUID-VkImageViewCreateInfo-imageViewFormatReinterpretation-04466`

I needed a way to check the `R, G, B` != `R, G, A,` and only way I could think of was a double for-loop

Comment (charles-lunarg) [reply include/vulkan/utility/vk_format_utils.h:2259]: Is this an example of a function which is useful for VVL and thats about it? Not sure how to use this function in not-validation code.

Comment (spencer-lunarg) [reply include/vulkan/utility/vk_format_utils.h:2259]: this is an example of where maybe the answer is 

"Have a function that returns `COMPONENT_INFO` and have the end users do the checking.... Like having `FormatsSameComponentBits` in here makes it easy to justify having so many other random util checks in as well that most people don't need.... I am not sure what level of abstraction for a Format Util header is useful, but this is on the edge of "very specific" use case IMO

Comment (charles-lunarg) [reply include/vulkan/utility/vk_format_utils.h:2259]: Since this is a header library, all those tools definitely exist. Before, it was all hidden in the .cpp file so users couldn't just get the info struct and compare the components themselves. But now? I would like to pursue removing this function and make VVL do what it needs (I'll of course implement that logic when I get around to making VVL use the format utils header).

Not sure what other functions in the header have a similar concern, as most everything takes a single format instead of 2 and most everything can't be easily checked themselves (lots of logic that benefits from codegen while this function does not).

File: tests/add_subdirectory/vk_format_utils.c (hunk 1 above)
Comment (juan-lunarg) [reply tests/add_subdirectory/vk_format_utils.c:8]: It not only tests the C code is valid.

It helps test add_subdirectory / find_package utilization.

Comment (juan-lunarg) [reply tests/add_subdirectory/vk_format_utils.c:8]: IE it's integration testing to make sure we are writing good CMake for users to absorb our libaries

File: tests/format_utils/CMakeLists.txt [hunk 9]
```diff
@@ -0,0 +1,25 @@
+# Copyright 2023 The Khronos Group Inc.
+# Copyright 2023 Valve Corporation
+# Copyright 2023 LunarG, Inc.
+#
+# SPDX-License-Identifier: Apache-2.0
+
+find_package(GTest REQUIRED CONFIG)
+find_package(magic_enum REQUIRED CONFIG)
+
+include(GoogleTest)
```
Comment (juan-lunarg) [inline tests/format_utils/CMakeLists.txt:10]: Note I don't like how this code is repeated in our CMake testing code. I'll make a fix for this in another PR though. It's currently consistent as is.

**Comment by juan-lunarg** [comment]
Did some minimal integration testing with VulkanTools. Looks good on that front 👍🏾 
https://github.com/LunarG/VulkanTools/pull/1883

File: include/vulkan/utility/vk_format_utils.h [hunk 10]
```diff
@@ -0,0 +1,2297 @@
...
+//     VK_IMAGE_ASPECT_PLANE_2_BIT -> 2
+//     <any other value> -> FORMAT_UTILS_MAX_PLANES
+inline uint32_t GetPlaneIndex(VkImageAspectFlagBits aspect);
+
+// Returns whether a format is of the numberical type SFLOAT
+// Format must only contain one numberical type, so formats like D16_UNORM_S8_UINT always return false
+inline bool FormatIsSFLOAT(VkFormat format);
+
+// Returns whether a format is of the numberical type SINT
+// Format must only contain one numberical type, so formats like D16_UNORM_S8_UINT always return false
```
Comment (juan-lunarg) [inline include/vulkan/utility/vk_format_utils.h:133]:
```suggestion
// Format must only contain one numerical type, so formats like D16_UNORM_S8_UINT always return false
```

File: include/vulkan/utility/vk_format_utils.h [hunk 11]
```diff
@@ -0,0 +1,2297 @@
...
+
+// Return the size in bytes of one texel of a foramt
+// For compressed or multi-plane, this may be a fractional number
+inline uint32_t FormatElementSizeWithAspect(VkFormat format, VkImageAspectFlagBits aspectMask);
+
+// Return the size in bytes of one texel of a foramt
+// Format must not be a depth, stencil, or multiplane format
+inline double FormatTexelSize(VkFormat format);
+
+// Return the size in bytes of one texel of a foramt
```
Comment (juan-lunarg) [inline include/vulkan/utility/vk_format_utils.h:278]:
```suggestion
// Return the size in bytes of one texel of a format
```

File: tests/format_utils/test_formats.cpp [hunk 12]
```diff
@@ -0,0 +1,681 @@
...
+    return comp_count;
+}
+
+TEST(FormatUtils, FormatIsSFLOAT) {
+    for (auto [format, format_str] : magic_enum::enum_entries<VkFormat>()) {
+        // special case depth + stencil formats
+        if (format == VK_FORMAT_D32_SFLOAT_S8_UINT) {
+            EXPECT_FALSE(FormatIsSFLOAT(format));
+            continue;
+        }
```
Comment (juan-lunarg) [inline tests/format_utils/test_formats.cpp:60]: Is the rational because the stencil isn't SFLOAT? IE this function only returns true if each component of the format is SFLOAT?

Comment (charles-lunarg) [reply tests/format_utils/test_formats.cpp:60]: This was how the function worked before, I do not know how rational such a decision is. I guess it makes sense, but I think asking who ever added the function originally would be the best person.

Comment (juan-lunarg) [reply tests/format_utils/test_formats.cpp:60]: It's documented so it's fine. Perhaps having a FormatHasSFLOAT would be appropriate to have. But that a later feature request for another PR.

Comment (spencer-lunarg) [reply tests/format_utils/test_formats.cpp:60]: I added this, there are some strange edge cases where one of these dual Depth/Stencil formats was returning true and we didn't want that behavior... I guess the `FormatHas` is just this but would return true for `D32S8` but not sure a use case... like if you are trying to match your Shader Interface, these dual formats are not allowed anyway

Comment (juan-lunarg) [reply tests/format_utils/test_formats.cpp:60]: >  but not sure a use case...

Right. I'm not sure there is a use case either. Which is why it's fine as is. If we need more functionality we can add it later.

File: include/vulkan/utility/vk_format_utils.h (hunk 2 above)
Comment (juan-lunarg) [reply include/vulkan/utility/vk_format_utils.h:35]: >  1. would it be better to forward declare so people looking at the file can quickly see the functions at the top?

I personally disagree on forward declaring. Most IDEs offer a way to collapse all functions/scopes. So IMO the forward declarations just adds more clutter/complexity.

Comment (juan-lunarg) [reply include/vulkan/utility/vk_format_utils.h:35]: > VK_ looks like a vulkan api enum which I do not want to be mistaken as.

FFTI. I think using `VUL_` would be more appropriate.

File: scripts/generators/format_utils_generator.py [hunk 13]
```diff
@@ -0,0 +1,695 @@
...
+#
+# SPDX-License-Identifier: Apache-2.0
+
+import os
+from generators.vulkan_object import (Format)
+from generators.base_generator import BaseGenerator
+
+# Make C++ name friendly class name
+def getClassName(className: str) -> str:
+    return className.replace('-', '').replace(' ', '_').upper()
```
Comment (juan-lunarg) [inline scripts/generators/format_utils_generator.py:15]: Is this needed? We are generating a C header.

Comment (charles-lunarg) [reply scripts/generators/format_utils_generator.py:15]: Oh absolutely - its 'C++ & C' friendly.

Spec uses `16-bit 2-plane 420` whereas we need `16BIT_2PLANE_420`.

File: include/vulkan/utility/vk_format_utils.h (hunk 7 above)
[the remark this answers is not in the document]
Comment (charles-lunarg) [reply include/vulkan/utility/vk_format_utils.h:166]: Let me know if the comments make more sense now.

File: include/vulkan/utility/vk_format_utils.h (hunk 3 above)
Comment (charles-lunarg) [reply include/vulkan/utility/vk_format_utils.h:127]: Added FORMAT_UTILS_INVALID_INDEX = 0xFFFFFFFF to make it clearer. The MAX_COMPONENTS is required since that creates the size of the components array.

File: include/vulkan/utility/vk_format_utils.h (hunk 2 above)
Comment (charles-lunarg) [reply include/vulkan/utility/vk_format_utils.h:36]: Forward declaring can be annoying since `enum` doesn't declare a size, but sizes are needed for forward declarations. I am not going to try to do that here, since I don't want to delay this PR any further while I wrestle getting enum forward decls working in C++ & C.

File: include/vulkan/utility/vk_format_utils.h [hunk 14]
```diff
@@ -0,0 +1,2268 @@
...
+
+inline bool FormatHasBlue(VkFormat format) { return FormatHasComponentType(format, FORMAT_UTILS_COMPONENT_TYPE_B); }
+
+inline bool FormatHasAlpha(VkFormat format) { return FormatHasComponentType(format, FORMAT_UTILS_COMPONENT_TYPE_A); }
+
+inline uint32_t GetPlaneIndex(VkImageAspectFlagBits aspect) {
+    switch (aspect) {
+        case VK_IMAGE_ASPECT_PLANE_0_BIT:
+            return 0;
+            break;
```
Comment (ziga-lunarg) [inline include/vulkan/utility/vk_format_utils.h:2253]: `break` after `return`?

Comment (charles-lunarg) [reply include/vulkan/utility/vk_format_utils.h]: Originally it didn't return from inside the switch statement, but now that it does, the break is redundant. Thanks for catching it!

**Review by juan-lunarg — CHANGES_REQUESTED** [review]
This will require updating `BUILD.gn`

File: include/vulkan/utility/vk_format_utils.h (hunk 2 above)
Comment (juan-lunarg) [reply include/vulkan/utility/vk_format_utils.h:36]: Agreed that's why I said FFTI.

**Review by juan-lunarg — APPROVED** [review]

**Comment by charles-lunarg** [comment]
Apologies for the force push spam - the C compatibility tests & add_subdirectory/find_package tests are easy to forget about since they aren't compiled normally.

The build.gn file should be correct now as well @juan-lunarg

**Review by juan-lunarg — APPROVED** [review]
LGTM
